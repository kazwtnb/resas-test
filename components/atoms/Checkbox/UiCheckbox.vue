<template>
  <div class="ui-checkbox" @click="onClick($event)">
    <label>
      <input type="checkbox" :checked="modelValue" />
      <div
        class="ui-checkbox-box"
        :class="{ 'ui-checkbox-box--checked': modelValue }"
      >
        <i class="fa-solid fa-check" />
      </div>
      <span>{{ label }}</span>
    </label>
  </div>
</template>
<script setup lang="ts">
interface Props {
  modelValue?: boolean;
  label: string;
}

interface Emits {
  (e: 'update:modelValue', value: boolean): void;
  (e: 'click', value: MouseEvent): void;
}

const props = withDefaults(defineProps<Props>(), {
  modelValue: false
});
const emit = defineEmits<Emits>();

const onClick = (event: MouseEvent): void => {
  event.preventDefault();
  emit('update:modelValue', !props.modelValue);
  emit('click', event);
};
</script>
<style lang="scss" scoped>
.ui-checkbox {
  display: inline-block;

  label {
    display: inline-flex;
    align-items: center;
    cursor: pointer;
  }

  input {
    border: 0;
    clip: rect(0 0 0 0);
    height: 1px;
    width: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    position: absolute;
    white-space: nowrap;
    pointer-events: none;
  }

  .ui-checkbox-box {
    display: inline-flex;
    width: 1rem;
    height: 1rem;
    align-items: center;
    justify-content: center;
    border: 1px solid colors.$color-border;
    border-radius: 2px;
    color: colors.$color-border;
    font-size: 0.75rem;
    transition: color 200ms ease-in, background-color 200ms ease-in;

    &:not(.ui-checkbox-box--checked) {
      > i {
        opacity: 0.3;
      }
    }

    &.ui-checkbox-box--checked {
      color: colors.$color-white;
      background-color: colors.$color-primary;
      border-color: colors.$color-primary;
    }

    + * {
      margin-left: 0.5rem;
    }
  }

  &:hover .ui-checkbox-box,
  input:focus-visible + .ui-checkbox-box {
    &:not(.ui-checkbox-box--checked) {
      background-color: colors.$color-hover;
    }

    &.ui-checkbox-box--checked {
      background-color: colors.$color-primary-hover;
      border-color: colors.$color-primary-hover;
    }
  }
}
</style>
