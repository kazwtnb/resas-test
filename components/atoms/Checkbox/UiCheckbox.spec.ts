import { mount } from '@vue/test-utils';
import { describe, expect, test } from 'vitest';
import UiCheckbox from './UiCheckbox.vue';

describe('UiCheckbox', () => {
  describe('click Event', () => {
    test.each([
      {
        name: 'チェックを入れる',
        input: {
          modelValue: false
        },
        expected: true
      },
      {
        name: 'チェックを外す',
        input: {
          modelValue: true
        },
        expected: false
      }
    ])(`%o`, ({ input, expected }) => {
      const wrapper = mount(UiCheckbox, {
        props: {
          modelValue: input.modelValue,
          label: 'value'
        }
      });

      wrapper.find('.ui-checkbox').trigger('click');

      expect(wrapper.emitted('update:modelValue')).toEqual([[expected]]);
      expect(wrapper.emitted('click')).toBeTruthy();
    });
  });

  describe('snapshot', () => {
    test.each([
      {
        name: '未選択',
        modelValue: false
      },
      {
        name: '選択済み',
        modelValue: true
      }
    ])(`%o`, (input) => {
      const wrapper = mount(UiCheckbox, {
        props: {
          modelValue: input.modelValue,
          label: 'value'
        }
      });
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
