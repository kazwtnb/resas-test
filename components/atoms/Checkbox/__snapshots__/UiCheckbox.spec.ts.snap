// Vitest Snapshot v1

exports[`UiCheckbox > snapshot > { name: '未選択', modelValue: false } 1`] = `
<div
  class="ui-checkbox"
  data-v-2d44444b=""
>
  <label
    data-v-2d44444b=""
  >
    <input
      data-v-2d44444b=""
      type="checkbox"
    />
    <div
      class="ui-checkbox-box"
      data-v-2d44444b=""
    >
      <i
        class="fa-solid fa-check"
        data-v-2d44444b=""
      />
    </div>
    <span
      data-v-2d44444b=""
    >
      value
    </span>
  </label>
</div>
`;

exports[`UiCheckbox > snapshot > { name: '選択済み', modelValue: true } 1`] = `
<div
  class="ui-checkbox"
  data-v-2d44444b=""
>
  <label
    data-v-2d44444b=""
  >
    <input
      data-v-2d44444b=""
      type="checkbox"
    />
    <div
      class="ui-checkbox-box ui-checkbox-box--checked"
      data-v-2d44444b=""
    >
      <i
        class="fa-solid fa-check"
        data-v-2d44444b=""
      />
    </div>
    <span
      data-v-2d44444b=""
    >
      value
    </span>
  </label>
</div>
`;
